doc/fundamentals/spouts.md

```diff
@@ -0,0 +1,291 @@
+`reader.Close()` won't be executed
+until after `main()` is finished.
+(This is a common idiom in Go;
+to `defer` the close of a resource
+right after you open it.)
+
+
+```Go
+	// And create a new kafka reader
+	reader := kafka.NewReader(kafka.ReaderConfig{
+		Brokers:  []string{host + ":" + port},
+		Topic:    topic,
+		MinBytes: 10e1,
+		MaxBytes: 10e6,
+	})
+	defer reader.Close()
+```
+
+### Open /pfs/out for writing
+
+We're opening the named pipe`/pfs/out` for writing
+so we can send it a `tar` archive stream
+with the file
+we want to output.
+Note that the named pipe has to be opened
+with write-only permissions.
+
+
+```Go
+    // Open the /pfs/out pipe with write only permissons (the pachyderm spout will be reading at the other end of this)
+	// Note: it won't work if you try to open this with read, or read/write permissions
+	out, err := os.OpenFile("/pfs/out", os.O_WRONLY, 0644)
+	if err != nil {
+		panic(err)
+	}
+	defer out.Close()
+```
+
+### Write the outer file loop
+Here we open a tar stream 
+into the directory we opened above,
+so that Pachyderm can place files
+into the output repo.
+For clarity's sake,
+we'll omit the message-processing loop
+inside this file loop.
+The `err` variable is used
+in the message-processing loop
+for errors reading from the stream
+and writing to the directory.
+The stream is opened 
+at the top of the loop
+and should be closed 
+at the bottom.
+In this case, 
+it'll be closed 
+after a message is processed.
+
+```Go
+	// this is the file loop
+	for {
+		if err := func() error {
+			tw := tar.NewWriter(out)
+			defer tw.Close()
+			// this is the message loop
+			for {
+
+            // ...omitted
+            
+			}
+		}(); err != nil {
+			panic(err)
+		}
+	}
+```
+
+### Create the message processing loop
+
+If you have trouble following this Go code,
+just read the text
+to get an idea of what you need to do.
+
+First, 
+we read a message
+from our Kafka queue.
+Note the use of a 5-second timeout
+on the read.
+
+```Go
+			// this is the message loop
+            for {
+				// read a message
+				ctx, cancel := context.WithTimeout(context.Background(), 5*time.Second)
+				defer cancel()
+				m, err := reader.ReadMessage(ctx)
+				if err != nil {
+					return nil
+                }
+```
+
+Then we create a filename
+and write that to the `tar` stream
+we opened at the beginning
+of the file loop.
+```Go
+                // give it a unique name
+				name := topic + time.Now().Format(time.RFC3339Nano)
+
+                // write the header
+				for err = tw.WriteHeader(&tar.Header{
+					Name: name,
+					Mode: 0600,
+					Size: int64(len(m.Value)),
+				}); err != nil; {
+					if !strings.Contains(err.Error(), "broken pipe") {
+						return err
+					}
+					// if there's a broken pipe, just give it some time to get ready for the next message
+					time.Sleep(5 * time.Millisecond)
+				}
+```
+
+Then we write the actual message 
+as the contents of the file.
+Note the use of a timeout
```

Comment:
This part is not right. This code does not miss any messages in this case -- it simply retries writing the message until it is successful or a more serious error occurs.

What's happening here is that the spout has its own loop where it opens the `pfs/out` pipe, reads form the tar stream, and then commits when the `pfs/out` pipe is closed. At this point, it needs to close and then reopen the `pfs/out` pipe itself. There can be a small delay here, in which case this code will try writing to the pipe before it has been reopened. This specific problem will cause the `broken pipe` error. And in this case, the correct thing to do is to wait a little bit and then try again, which is what we are doing here. For robustness, we repeatedly try until we either succeed or get a different error -- this accounts for variability in the timing on the spout end.

FWIW, in the case that we get a more serious error, it is typically going to be because the spout itself has crashed on the pachyderm side. That error will appear in the logs for the pipeline. Additionally, the panic'd errors here will be visible in the pipeline logs as well, prefaced with `error running user code:`.

Reply:
Also worth noting that if we batching the messages in time intervals, then for sufficiently large time intervals (say 20ms), then doing this will not be necessary, since the spout should have had plenty of time to reopen the pipe in between commits.

Reply:
This was very helpful!  Thank you for correcting my misreading of the code.  I've updated the text with this points.